### PR TITLE
[Merged by Bors] - feat: port Order.GaloisConnection

### PR DESCRIPTION
**Pull request by dupuisf**

### CODE REVIEW

File: Mathlib/Order/GaloisConnection.lean [hunk 1]
```diff
@@ -0,0 +1,926 @@
...
+
+section lift
+
+variable [PartialOrder α]
+
+-- See note [reducible non instances]
+/-- Lift the infima along a Galois coinsertion -/
+@[reducible]
+def liftSemilatticeInf [SemilatticeInf β] (gi : GaloisCoinsertion l u) : SemilatticeInf α :=
+  { ‹PartialOrder α› with
```
Comment (dupuisf) [inline Mathlib/Order/GaloisConnection.lean:858]: In this and the following few defs, the elaborator struggled with `αᵒᵈ` vs `α`; now it compiles but *much* slower than in mathlib3.

File: Mathlib/Order/GaloisConnection.lean [hunk 2]
```diff
@@ -0,0 +1,934 @@
...
+theorem isLUB_of_l_image [Preorder α] [Preorder β] (gi : GaloisCoinsertion l u) {s : Set α} {a : β}
+    (hs : IsLUB (l '' s) a) : IsLUB s (u a) :=
+  gi.dual.isGLB_of_u_image hs
+#align galois_coinsertion.is_lub_of_l_image GaloisCoinsertion.isLUB_of_l_image
+
+section lift
+
+variable [PartialOrder α]
+
+-- See note [reducible non instances]
```
Comment (semorrison) [inline Mathlib/Order/GaloisConnection.lean:862]:
```suggestion
-- Porting note: In this and the following few defs, the elaborator struggled with αᵒᵈ vs α;
-- now it compiles but much slower than in mathlib3.

-- See note [reducible non instances]
```

File: Mathlib/Order/GaloisConnection.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Order/GaloisConnection.lean:869]: I've left a porting note about this.

File: Mathlib/Order/GaloisConnection.lean [hunk 3]
```diff
@@ -0,0 +1,937 @@
...
+    (hs : IsLUB (l '' s) a) : IsLUB s (u a) :=
+  gi.dual.isGLB_of_u_image hs
+#align galois_coinsertion.is_lub_of_l_image GaloisCoinsertion.isLUB_of_l_image
+
+section lift
+
+variable [PartialOrder α]
+
+-- Porting note: In this and the following few defs, the elaborator struggled with αᵒᵈ vs α;
+-- now it compiles but much slower than in mathlib3.
```
Comment (semorrison) [inline Mathlib/Order/GaloisConnection.lean:863]:
```suggestion
-- now it compiles but much slower than in mathlib3.
-- Possibly related to the issue discussed at
-- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Performance.20issue.20with.20.60CompleteBooleanAlgebra.60/near/316760798
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**